File: src/utils/precipitation.py
import pandas
import math
import json

weather = pandas.read_csv('F:\MM-804\src\data\climate-daily.csv')

rain = {}
snow = {}

for i, row in weather.iterrows():
    year = row['LOCAL_DATE'][:4]
    
    if not math.isnan(row['TOTAL_RAIN']):
        rain[year] = rain.get(year, []) + [row['TOTAL_RAIN']]
    else:
        rain[year] = rain.get(year, []) + [0]

    if not math.isnan(row['TOTAL_SNOW']):
        snow[year] = snow.get(year, []) + [row['TOTAL_SNOW']]
    else:
        snow[year] = snow.get(year, []) + [0]

# years = []
# for keys in rain.keys():
#     years.append(int(keys))

rain_data = []
for values in rain.values():
    rain_data.append(values)

snow_data = []
for values in snow.values():
    snow_data.append(values)

with open('F:/MM-804/src/data/rain.json', 'w') as output_file:
    json.dump(rain_data, output_file)
with open('F:/MM-804/src/data/snow.json', 'w') as output_file:
    json.dump(snow_data, output_file)

# with open('../data/years.json', 'w') as output_file:
#     json.dump(years, output_file)

File: src/components/Precipitation.js
import React, { useState, useEffect } from 'react';
import Plot from 'react-plotly.js';
import Papa from 'papaparse';
import groupBy from 'lodash/groupBy';

import precipitationData from '../data/climate-daily.csv';

function Precipitation() {
  const [data, setData] = useState([]);
  const [selectedYear, setSelectedYear] = useState(null);

  useEffect(() => {
    const fetchParseData = async () => {
      Papa.parse(precipitationData, {
        download: true,
        header: true,
        delimiter: ",",
        complete: ((result) => {
          setData(result.data);
        })
      })
    }
    fetchParseData();
  }, []);

  const groupByTimestamp = groupBy(data, 'LOCAL_DATE');
  const dates = Object.keys(groupByTimestamp)
    .filter(date => {
      const counts = groupByTimestamp[date];
      return counts.some(count => parseFloat(count.TOTAL_RAIN) >= 0 || parseFloat(count.TOTAL_SNOW) >= 0);
    })
    .filter((date) => !selectedYear || new Date(date).getFullYear() === selectedYear)
    .sort((a, b) => new Date(a) - new Date(b));

  const rain = dates.map((date) => {
    const counts = groupByTimestamp[date];
    return counts.reduce((total, count) => total + parseFloat(count.TOTAL_RAIN), 0);
  });

  const snow = dates.map((date) => {
    const counts = groupByTimestamp[date];
    return counts.reduce((total, count) => total + parseFloat(count.TOTAL_SNOW), 0)*10;
  });

  const years = Array.from(new Set(data.map((row) => new Date(row.LOCAL_DATE).getFullYear())))
  .filter(year => !isNaN(parseInt(year)));

  const traces = [
    {
      x: dates,
      y: rain,
      type: 'bar',
      name: 'Rain',
      barmode: 'stack',
      marker: { color: '#448' }
    },
    {
      x: dates,
      y: snow,
      type: 'bar',
      name: 'Snow',
      barmode: 'stack',
      marker: { color: '#88C' }
    },
  ];

  return (
    <div>      
      <Plot
        data={traces}
        layout={{
          title: selectedYear ? `Precipitation Distribution for Year ${selectedYear}` : 'Precipitation Distribution',
          xaxis: { title: 'Date' },
          yaxis: { title: 'Precipitation (mm)' },
          width: 800,
          height: 600,
          autosize: true,
          barmode: 'stack',
        }}
      />
      <div style={{ marginTop: '20px', marginBottom: '80px' }}>
        <label htmlFor="year-dropdown">Select Year: </label>
        <select id="year-dropdown" value={selectedYear} onChange={(e) => setSelectedYear(parseInt(e.target.value))}>
          <option value={null}>All Years</option>
          {years.map((year) => (
            <option key={year} value={year}>{year}</option>
          ))}
        </select>
      </div>
    </div>
  );
}

export default Precipitation;
